File: Python/add_time_zone.py
import pandas as pd
import pytz

def add_time_zone(datetime_series, tz_series):
    """
    Add Time Zone to Datetime Objects.
    
    This function takes a pandas Series of datetime objects and a corresponding Series of time zone strings,
    and returns a new datetime Series where each datetime object has been assigned the corresponding time zone.
    
    :param datetime_series: pd.Series of pd.Timestamp, representing the datetime objects to which the time zones will be added.
    :param tz_series: pd.Series of str, representing the time zones to be assigned to the datetime objects.
                     The length of tz_series should be the same as the length of datetime_series.
    :return: A pd.Series of pd.Timestamp with the same length as the input, where each datetime object has the assigned time zone.
    :raise ValueError: If the length of datetime_series and tz_series are not the same.
    """
    
    if len(datetime_series) != len(tz_series):
        raise ValueError("Length of datetime_series and tz_series must be the same.")
    
    # Initializing the new_datetime series
    new_datetime = pd.Series(index=datetime_series.index, dtype='datetime64[ns]')
    
    for i in range(len(datetime_series)):
        tz = pytz.timezone(tz_series.iloc[i])
        new_datetime.iloc[i] = datetime_series.iloc[i].tz_convert(tz)
    
    return new_datetime

# Define a DataFrame
df = pd.DataFrame({
    'datetime': pd.to_datetime(["2023-09-23 12:00:00", "2023-09-24 12:00:00", "2023-09-25 12:00:00"], utc=True)
})

# Define a Series of time zones
tz_series = pd.Series(["America/New_York", "Europe/London", "Asia/Tokyo"])

# Apply the add_timezone function to the datetime column with different time zones for each element
df['new_datetime'] = add_time_zone(df['datetime'], tz_series)

# Print the resulting DataFrame
print(df)

File: R/add_time_zone.R
# Define the enhanced add_time_zone function
add_time_zone <- function(datetime, tz) {
  #' Add Time Zone to Datetime Objects
  #'
  #' This function takes a vector of datetime objects and a corresponding vector of time zone strings,
  #' and returns a new datetime vector where each datetime object has been assigned the corresponding time zone.
  #'
  #' @param datetime POSIXct vector, representing the datetime objects to which the time zones will be added.
  #' @param tz Character vector, representing the time zones to be assigned to the datetime objects.
  #'           The length of tz should be the same as the length of datetime.
  #' @return A POSIXct vector with the same length as the input, where each datetime object has the assigned time zone.
  if(length(datetime) != length(tz)) {
    stop("Length of datetime vector and tz vector must be the same.")
  }
  
  # Loop over each element of the datetime vector and assign the corresponding timezone
  new_datetime <- vector("list", length(datetime))
  for(i in seq_along(datetime)) {
    new_datetime[[i]] <- as.POSIXct(format(datetime[i], tz = tz[i]))
  }
  
  # Convert the list to a POSIXct object
  new_datetime <- do.call(c, new_datetime)
  return(new_datetime)
}

# Define a dataframe
df <- data.frame(
  datetime = as.POSIXct(c("2023-09-23 12:00:00", "2023-09-24 12:00:00", "2023-09-25 12:00:00"), tz = "UTC")
)

# Define a vector of time zones
tz_vector <- c("America/New_York", "Europe/London", "Asia/Tokyo")

# Apply the add_time_zone function to the datetime column with different time zones for each element
df$new_datetime <- add_time_zone(df$datetime, tz_vector)

# Print the resulting dataframe
print(df)
